src/persist/README.md

```diff
@@ -3,49 +3,194 @@
+  used to allow read-only "tail"ing of an active persisted stream. Internally,
+  this stream name gets transparently mapped to a smaller internal stream id.
+- _blob_: A durable `&str` -> `[]byte` key value store. Persistence is
+  written in terms of a generic `Blob` and `Buffer` interface. Examples of blob
+  implementations include files in a folder and S3 (or other cloud storage).
+- _buffer_: A durable log with `[]byte` entries. Persistence is written in terms
+  of a generic `Blob` and `Buffer` interface. Examples of buffer implementations
+  include a traditional write-ahead log (possibly on EBS or similar) and Kafka.
+  TODO: Rename this to something better. Log?
+- _runtime_: The internally managed threads, doing the work of persisting new
+  records, compacting previously written ones, and managing read handles to
+  persistent data. Accessed externally via a thread-safe and clone-able
+  "client".
 
-### V1 and V2
+## External API
 
-`Persister` corresponds to [`PersisterV1` in the prototype][persister v1] and
-basically just supports synchronously writing data out, grabbing a snapshot of
-everything written so far, and unblocking compaction of everything less than a
-given timestamp. It could implemented on top of a SQLite table with something
-like the following schema:
+A persistence user initializes a `Buffer` and a `Blob` implementation and uses
+them to start the runtime, receiving a `RuntimeClient` in response. This client
+is the primary external interface to the persistence system. This runtime allows
+us to funnel everything a process is persisting through a single place for
+batching and rate limiting. The buffer and blob are exclusively locked for
+writes by this process (they can be opened for reads by other processes,
+however).
 
-[persister v1]: https://github.com/danhhz/differential-dataflow/blob/02673114b05933341893ab603327237a9583e432/persist/src/persister.rs#L30-L39
+The runtime stops when every copy of the client has been dropped (or when any
+client's stop method has been called), attempting to first gracefully finish any
+pending work, but rejecting new work.
 
-```sql
-CREATE TABLE persisted (persisted_id INT, key BYTES, val BYTES, ts BYTES, diff BYTES);
-CREATE INDEX ON persisted (persisted_id, ts, key);
+As needed, the user creates a persisted stream by name, returning a token. At
+dataflow construction time, this token can be exchanged to create one of several
+available persistence operators (see below). When the operator is run, it hands
+any blocking io or cpu heavy work to the runtime, asking the dataflow to be
+scheduled again when this finishes. If the same buffer and blob implementations
+were previously used to persist a stream with the same name, this previous data
+is read and also output by the operator, yielding periodically to allow
+downstream operators a chance to process it.
+
+## Operator Guarantees
+
+- _idempotence_: Any record emitted by a persistence operator is guaranteed to
```

Comment:
Nit, but this reads like *durability* rather than *idempotence*. Once a record is observed in the output, it is certain to be observed in future outputs of "the same thing". Is there an explainer for why idempotence is the right term here?

Reply:
I see; I def did use idempotence, with the intent meaning "if someone runs multiple copies of a dataflow, the rest of the world gets the same effect as if it ran just one". So, e.g. if the dataflow is going to write out transactions to Kafka, we would need it to check before doing so to avoid duplicate transactions. If persistence does something else as part of persisting the source, we would want it to not imperil correctness if it happens twice (or, we might not care at the moment, but want to aim at idempotence).

Reply:
I think this is covered by "definite", now that we have that word